planner/core/logical_plan_builder.go

```diff
@@ -2397,6 +2395,31 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
 		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.ShowViewPriv, dbName.L, tableInfo.Name.L, "", ErrViewNoExplain)
 	}
 
+	if tableInfo.View.Cols == nil {
```

Comment:
Can we extract the two similar code block to a function?

Reply:
It's not easy, how about this way?

Reply:
We can extract 2042~2434 to a function like buildProjUponView.

Reply:
Like this:
``` go
func (b *PlanBuilder) BuildProjUponView(ctx context.Context, dbName model.CIStr, tableInfo *model.TableInfo, selectLogicalPlan Plan) (Plan, error){
	columnInfo := tableInfo.Cols()
	cols := selectLogicalPlan.Schema().Columns
	// In the old version of VIEW implementation, tableInfo.View.Cols is used to
	// store the origin columns' names of the underlying SelectStmt used when
	// creating the view.
	if tableInfo.View.Cols != nil {
		cols = cols[:0]
		for _, info := range columnInfo {
			col := selectLogicalPlan.Schema().FindColumnByName(info.Name.L)
			if col == nil {
				return nil, ErrViewInvalid.GenWithStackByArgs(dbName.O, tableInfo.Name.O)
			}
			cols = append(cols, col)
		}
	}
	projSchema := expression.NewSchema(make([]*expression.Column, 0, len(tableInfo.Columns))...)
	projExprs := make([]expression.Expression, 0, len(tableInfo.Columns))
	for i, col := range cols {
		origColName := col.ColName
		if tableInfo.View.Cols != nil {
			origColName = tableInfo.View.Cols[i]
		}
		projSchema.Append(&expression.Column{
			UniqueID:    b.ctx.GetSessionVars().AllocPlanColumnID(),
			TblName:     col.TblName,
			OrigTblName: col.OrigTblName,
			ColName:     columnInfo[i].Name,
			OrigColName: origColName,
			DBName:      col.DBName,
			RetType:     col.GetType(),
		})
		projExprs = append(projExprs, col)
	} 
	projUponView := LogicalProjection{Exprs: projExprs}.Init(b.ctx)
	projUponView.SetChildren(selectLogicalPlan.(LogicalPlan))
	projUponView.SetSchema(projSchema)
	return projUponView, nil
}

```

Reply:
Thanks

Reply:
@XuHuaiyu PTAL again